### PR TITLE
rostest: use AnyMsg in publishtest

### PR DESCRIPTION
**Pull request by beetleskin**
Right now publishtest only works correctly when a publisher of the given topic exists: If no publisher exists, the retrieval of the topic class blocks until the test times out. This leads to aborted tests when one would expect a failed one with correct error message, stating that no message on the given topic was published. Using rospy.AnyMsg fixes that.

This also makes it impossible to test that a certain topic is not being published, when no publisher exists.

### CODE REVIEW

File: tools/rostest/CMakeLists.txt [hunk 1]
```diff
@@ -6,18 +6,24 @@ find_package(Boost COMPONENTS system thread)
...
-install(DIRECTORY include/${PROJECT_NAME}/
+  DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/nodes
+)
+
+install(
+  DIRECTORY include/${PROJECT_NAME}/
   DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
-  FILES_MATCHING PATTERN "*.h")
+  FILES_MATCHING PATTERN "*.h"
+)
```
Comment (dirk-thomas) [inline tools/rostest/CMakeLists.txt]: Please revert all of the above unrelated whitespace changes. They just make backporting patches between branches unnecessarily more difficult.

File: tools/rostest/test/publishtest.test [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+                timeout: 2.
+              - name: /once_topic
+                timeout: 2.
+              - name: /unpublished_topic
+                timeout: 2.
+                negative: true
+        </rosparam>
+    </test>
+
+</launch>
```
Comment (dirk-thomas) [inline tools/rostest/test/publishtest.test]: Please add a trailing newline.

File: tools/rostest/test/hztest.test [hunk 3]
```diff
@@ -6,19 +6,6 @@
...
-    <rosparam>
-      topic: chatter
-      hz: 10.0
-      hzerror: 0.5
-      test_duration: 5.0
-      wait_time: 21.0
-    </rosparam>
-  </test>
-
-  -->
```
Comment (dirk-thomas) [inline tools/rostest/test/hztest.test]: Why should this commented block be removed?

**Review by dirk-thomas — COMMENTED** [review]
Please create separate PRs for unrelated changes: one for the `rospy.AnyMsg` change and one for the `retry` flag to address the flakiness.

File: tools/rostest/CMakeLists.txt (hunk 1 above)
Comment (beetleskin) [reply tools/rostest/CMakeLists.txt]: but .. its just messy in there

File: tools/rostest/test/hztest.test (hunk 3 above)
Comment (beetleskin) [reply tools/rostest/test/hztest.test]: why should it be in there?

File: tools/rostest/test/publishtest.test (hunk 2 above)
Comment (beetleskin) [reply tools/rostest/test/publishtest.test]: there is a trailing newline in there ..

File: tools/rostest/test/hztest.test (hunk 3 above)
Comment (dirk-thomas) [reply tools/rostest/test/hztest.test]: Same reason as above:

> Please revert all ... unrelated ... changes. They just make backporting patches between branches unnecessarily more difficult.

**Comment by beetleskin** [comment]
Done, anything missing?

**Comment by dirk-thomas** [comment]
LGTM. Thanks for the patch.

**Comment by beetleskin** [comment]
any plans to release this?

**Comment by dirk-thomas** [comment]
> any plans to release this?

See #1802.
